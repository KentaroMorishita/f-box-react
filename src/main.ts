export * from "./useBox";
export * from "./useRBox";
export * from "./useRBoxForm";
export * from "./useRBoxResource";
export * from "./useRBoxTransaction";
