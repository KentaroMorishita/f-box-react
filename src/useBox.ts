import { useMemo } from "react";
import { Box } from "f-box-core";

export function useBox<T>(source: Box<T>): [T, Box<T>];
export function useBox<T>(
  source: T | (() => T | Box<T>),
  deps?: React.DependencyList
): [T, Box<T>];
export function useBox<T>(
  source: T | Box<T> | (() => T | Box<T>),
  deps: React.DependencyList = []
): [T, Box<T>] {
  const box = useMemo<Box<T>>(() => {
    const value =
      typeof source === "function" ? (source as () => T | Box<T>)() : source;
    return Box.isBox(value) ? (value as Box<T>) : Box.pack(value as T);
  }, deps);

  const value = box.getValue();

  return [value, box];
}
