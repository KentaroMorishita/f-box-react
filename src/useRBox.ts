import { useMemo } from "react";
import { useSyncExternalStore } from "react";
import { RBox } from "f-box-core";

export function useRBox<T>(source: RBox<T>): [T, RBox<T>];
export function useRBox<T>(
  source: T | (() => T | RBox<T>),
  deps?: React.DependencyList
): [T, RBox<T>];
export function useRBox<T>(
  source: T | RBox<T> | (() => T | RBox<T>),
  deps: React.DependencyList = []
): [T, RBox<T>] {
  const box = useMemo<RBox<T>>(() => {
    const value =
      typeof source === "function" ? (source as () => T | RBox<T>)() : source;
    return RBox.isRBox(value) ? (value as RBox<T>) : RBox.pack(value as T);
  }, deps);

  const value = useSyncExternalStore(
    (onStoreChange) => {
      const key = box.subscribe(onStoreChange);
      return () => box.unsubscribe(key);
    },
    () => box.getValue()
  );

  return [value, box];
}

export const set = RBox.set;
