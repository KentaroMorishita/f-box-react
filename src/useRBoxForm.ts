import { useRBox } from "./useRBox";

type FormValues = Record<string, unknown>;

type ValidationRule = () => boolean;

type Validation<T extends FormValues> = {
  [K in keyof T]: ValidationRule[];
};

type Edited<T extends FormValues> = {
  [K in keyof T]: boolean;
};

type UseRBoxFormResult<T extends FormValues> = {
  form: T;
  handleChange: <K extends keyof T>(field: K, value: T[K]) => void;
  handleValidatedSubmit: (
    onSuccess: (form: T) => void,
    onError?: () => void
  ) => (e: React.FormEvent) => void;
  shouldShowError: <K extends keyof T>(field: K) => (index: number) => boolean;
  validation: Validation<T>;
  edited: Edited<T>;
  formValid: boolean;
  resetForm: () => void;
  markAllEdited: () => void;
};

export function useRBoxForm<T extends FormValues>(
  initialValues: T,
  validate: (form: T) => Validation<T>
): UseRBoxFormResult<T> {
  const initialEdited = (value: boolean): Edited<T> =>
    Object.keys(initialValues).reduce(
      (acc, key) => ({ ...acc, [key]: value }),
      {} as Edited<T>
    );

  // RBoxes
  const [form, formBox] = useRBox<T>(initialValues);
  const [validation, validationBox] = useRBox(() => formBox["<$>"](validate));
  const [edited, editedBox] = useRBox<Edited<T>>(initialEdited(false));
  const [formValid, formValidBox] = useRBox(() =>
    validationBox["<$>"]((v) =>
      Object.values(v)
        .flat()
        .every((rule) => rule())
    )
  );

  // helpers
  const handleChange = <K extends keyof T>(field: K, value: T[K]) => {
    formBox.setValue((prev) => ({ ...prev, [field]: value }));
    editedBox.setValue((prev) => ({ ...prev, [field]: true }));
  };

  const handleValidatedSubmit =
    (onSuccess: (form: T) => void, onError?: () => void) =>
    (e: React.FormEvent) => {
      e.preventDefault();
      markAllEdited();

      if (!formValidBox.getValue()) {
        if (onError) onError();
        return;
      }
      onSuccess(formBox.getValue());
    };

  const shouldShowError =
    <K extends keyof T>(field: K) =>
    (index: number): boolean =>
      edited[field] && !validation[field][index]();

  const resetForm = () => {
    formBox.setValue(() => initialValues);
    editedBox.setValue(() => initialEdited(false));
  };

  const markAllEdited = () => {
    editedBox.setValue((prev) =>
      Object.keys(prev).reduce(
        (acc, key) => ({ ...acc, [key]: true }),
        {} as Edited<T>
      )
    );
  };

  return {
    form,
    handleChange,
    handleValidatedSubmit,
    shouldShowError,
    validation,
    edited,
    formValid,
    resetForm,
    markAllEdited,
  };
}
