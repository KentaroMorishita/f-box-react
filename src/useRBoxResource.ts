import { useRBox, set } from "./useRBox";
import { Either, Task, RBox } from "f-box-core";

const { right, left } = Either;

type Result<T> = Either<Error, Awaited<T>>;
type CacheMap<T> = Map<string, Result<T>>;

type RBoxResourceOptions<T> = {
  resultBox: RBox<Result<T>>;
  isLoadingBox: RBox<boolean>;
  cacheBox: RBox<CacheMap<T>>;
  isAutoRun: boolean;
  maxCacheSize: number;
};

type ResourceController<A> = {
  params: A;
  run: () => void;
  mutate: (args: Partial<A>) => void;
  refetch: () => void;
  clearCache: () => void;
};

const DEFAULT_IS_AUTO_RUN = true;
const DEFAULT_MAX_CACHE_SIZE = 100;
const DEFAULT_ERROR_OBJECT = new Error("No data yet");

const hash = (value: unknown, length: number = 16) => {
  const json = JSON.stringify(value);
  return Task.tryCatch<string>(
    async () => {
      if (typeof window !== "undefined" && window.crypto?.subtle) {
        const encoder = new TextEncoder();
        const data = encoder.encode(json);
        const hashBuffer = await crypto.subtle.digest("SHA-256", data);
        const hashArray = Array.from(new Uint8Array(hashBuffer));
        const fullHash = hashArray
          .map((b) => b.toString(16).padStart(2, "0"))
          .join("");
        return fullHash.slice(0, length);
      } else {
        const { createHash } = await import("crypto");
        const fullHash = createHash("sha256").update(json).digest("hex");
        return fullHash.slice(0, length);
      }
    },
    (err) => {
      console.error(err);
      return json;
    }
  );
};

export const castError = (err: unknown) =>
  err instanceof Error ? err : new Error(String(err));

export function useRBoxResource<T, A extends {}>(
  fetcher: (args: A) => Task<T>,
  initialParams: A = {} as A,
  options: Partial<RBoxResourceOptions<T>> = {}
): [Result<T>, boolean, ResourceController<A>] {
  const [result, resultBox] = useRBox<Result<T>>(
    () => options.resultBox ?? RBox.pack<Result<T>>(left(DEFAULT_ERROR_OBJECT))
  );
  const [isLoading, isLoadingBox] = useRBox(
    () => options.isLoadingBox ?? RBox.pack(false)
  );
  const [, cacheBox] = useRBox(
    () => options.cacheBox ?? RBox.pack<CacheMap<T>>(new Map())
  );

  const isAutoRun = options.isAutoRun ?? DEFAULT_IS_AUTO_RUN;
  const maxCacheSize = options.maxCacheSize ?? DEFAULT_MAX_CACHE_SIZE;
  const setResult = set(resultBox);

  const [params, paramsBox] = useRBox(initialParams);
  const mutate = (args: Partial<A>) =>
    paramsBox.setValue((prev) => ({ ...prev, ...args }));

  const runTask = async (currentParams: A, ignoreCache = false) => {
    const cacheKey = await hash(currentParams).run();
    const currentCache = cacheBox.getValue();
    const setCache = (result: Result<T>) =>
      cacheBox.setValue((prev) => {
        const updatedCache = new Map(prev);
        updatedCache.set(cacheKey, result);

        if (updatedCache.size > maxCacheSize) {
          const oldestKey = updatedCache.keys().next().value;
          if (oldestKey) updatedCache.delete(oldestKey);
        }
        return updatedCache;
      });

    if (!ignoreCache && currentCache.has(cacheKey)) {
      const cacheValue = currentCache.get(cacheKey)!;
      setResult(cacheValue);
      return;
    }

    isLoadingBox.setValue(true);

    Task.tryCatch<Result<T>>(
      async () => right(await fetcher(currentParams).run()),
      (err) => left(castError(err))
    )
      ["<$>"]((result) => {
        setResult(result);
        setCache(result);
      })
      .run()
      .finally(() => isLoadingBox.setValue(false));
  };

  const run = () => paramsBox["<$>"](runTask);
  const refetch = () => runTask(params, true);
  const clearCache = () => cacheBox.setValue(new Map());

  if (isAutoRun) {
    useRBox(run);
  }

  return [result, isLoading, { params, run, mutate, refetch, clearCache }];
}
