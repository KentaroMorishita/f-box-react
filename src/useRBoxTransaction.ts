import { RBox } from "f-box-core";
import { useRBox, set } from "./useRBox";

export type UseRBoxTransactionResult = [
  boolean,
  (action: () => Promise<void>) => Promise<void>
];

export function useRBoxTransaction(
  isPengingBox: RBox<boolean>
): UseRBoxTransactionResult {
  const [isPending] = useRBox(isPengingBox);
  const setIsPending = set(isPengingBox);
  const startTransaction = async (action: () => Promise<void>) => {
    setIsPending(true);
    await action();
    setIsPending(false);
  };

  return [isPending, startTransaction];
}
