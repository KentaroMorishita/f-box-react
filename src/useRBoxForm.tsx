import * as React from "react";
import { useRBox } from "./useRBox";
import { useRBoxTransaction } from "./useRBoxTransaction";

type FormValues = Record<string, unknown>;

type ValidationRule = () => boolean;

type Validation<T extends FormValues> = {
  [K in keyof T]: ValidationRule[];
};

type Edited<T extends FormValues> = {
  [K in keyof T]: boolean;
};

type UseRBoxFormResult<T extends FormValues> = {
  form: T;
  isPending: boolean;
  handleChange: <K extends keyof T>(field: K, value: T[K]) => void;
  handleValidatedSubmit: (
    onSuccess: (form: T) => Promise<void>,
    onError?: () => void
  ) => (e: React.FormEvent) => void;
  shouldShowError: <K extends keyof T>(field: K) => (index: number) => boolean;
  renderErrorMessages: <K extends keyof T>(
    field: K,
    messages: string[],
    Component?: React.ElementType
  ) => React.ReactElement;
  validation: Validation<T>;
  edited: Edited<T>;
  formValid: boolean;
  resetForm: () => void;
  markAllEdited: () => void;
};

export function useRBoxForm<T extends FormValues>(
  initialValues: T,
  validate: (form: T) => Validation<T>,
  ErrorComponent: React.ElementType = "span"
): UseRBoxFormResult<T> {
  const initialEdited = (value: boolean): Edited<T> =>
    Object.keys(initialValues).reduce(
      (acc, key) => ({ ...acc, [key]: value }),
      {} as Edited<T>
    );

  // RBoxes
  const [form, formBox] = useRBox<T>(initialValues);
  const [validation, validationBox] = useRBox(() => formBox["<$>"](validate));
  const [edited, editedBox] = useRBox<Edited<T>>(initialEdited(false));
  const [formValid, formValidBox] = useRBox(() =>
    validationBox["<$>"]((v) =>
      Object.values(v)
        .flat()
        .every((rule) => rule())
    )
  );

  const [, isPendingBox] = useRBox(false);
  const [isPending, startTransaction] = useRBoxTransaction(isPendingBox);

  // helpers
  const handleChange = <K extends keyof T>(field: K, value: T[K]) => {
    formBox.setValue((prev) => ({ ...prev, [field]: value }));
    editedBox.setValue((prev) => ({ ...prev, [field]: true }));
  };

  const handleValidatedSubmit =
    (onSuccess: (form: T) => Promise<void>, onError?: () => void) =>
    (e: React.FormEvent) => {
      e.preventDefault();
      markAllEdited();
      if (!formValidBox.getValue()) {
        if (onError) onError();
        return;
      }

      startTransaction(async () => onSuccess(formBox.getValue()));
    };

  const shouldShowError =
    <K extends keyof T>(field: K) =>
    (index: number): boolean =>
      edited[field] && !validation[field][index]();

  const renderErrorMessages = <K extends keyof T>(
    field: K,
    messages: string[],
    Component: React.ElementType = ErrorComponent
  ): React.ReactElement => (
    <>
      {messages.map(
        (message, index) =>
          shouldShowError(field)(index) && (
            <Component key={index}>{message}</Component>
          )
      )}
    </>
  );

  const resetForm = () => {
    formBox.setValue(() => initialValues);
    editedBox.setValue(() => initialEdited(false));
  };

  const markAllEdited = () => {
    editedBox.setValue((prev) =>
      Object.keys(prev).reduce(
        (acc, key) => ({ ...acc, [key]: true }),
        {} as Edited<T>
      )
    );
  };

  return {
    form,
    isPending,
    handleChange,
    handleValidatedSubmit,
    shouldShowError,
    renderErrorMessages,
    validation,
    edited,
    formValid,
    resetForm,
    markAllEdited,
  };
}
