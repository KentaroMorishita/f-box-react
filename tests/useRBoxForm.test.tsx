import { renderHook, act } from "@testing-library/react";
import { useRBoxForm } from "../src/useRBoxForm";

type Form = {
  name: string;
  age: number;
  isChecked: boolean;
};

const initialValues: Form = {
  name: "",
  age: 0,
  isChecked: false,
};

const validate = (form: Form) => ({
  name: [
    () => form.name.trim().length >= 3,
    () => /^[a-zA-Z]+$/.test(form.name),
  ],
  age: [() => form.age > 18],
  isChecked: [() => form.isChecked === true],
});

describe("useRBoxForm", () => {
  test("should initialize with initial values and default state", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    const { form, validation, edited } = result.current;

    expect(form).toEqual(initialValues);

    // validationを評価して結果を比較
    expect(
      Object.fromEntries(
        Object.entries(validation).map(([key, rules]) => [
          key,
          rules.map((rule) => rule()),
        ])
      )
    ).toEqual({
      name: [false, false],
      age: [false],
      isChecked: [false],
    });

    expect(edited).toEqual({
      name: false,
      age: false,
      isChecked: false,
    });
  });

  test("should update form values and mark field as edited on handleChange", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    act(() => {
      result.current.handleChange("name", "John");
    });

    const { form, edited, validation } = result.current;

    expect(form.name).toBe("John");
    expect(edited.name).toBe(true);

    // validationを評価して結果を比較
    expect(validation.name.map((rule) => rule())).toEqual([true, true]); // Name "John" satisfies both rules
  });

  test("should validate correctly for invalid inputs", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    act(() => {
      result.current.handleChange("name", "Jo"); // Too short, fails rule 1
    });

    const { validation, shouldShowError } = result.current;

    // validationを評価して結果を比較
    expect(validation.name.map((rule) => rule())).toEqual([false, true]); // Fails length rule, passes regex

    // shouldShowErrorを使った検証
    expect(shouldShowError("name")(0)).toBe(true); // Rule 0 should fail
    expect(shouldShowError("name")(1)).toBe(false); // Rule 1 should pass
  });

  test("should reset form values and edited state on resetForm", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    act(() => {
      result.current.handleChange("name", "John");
      result.current.handleChange("age", 25);
      result.current.resetForm();
    });

    const { form, edited } = result.current;

    expect(form).toEqual(initialValues);
    expect(edited).toEqual({
      name: false,
      age: false,
      isChecked: false,
    });
  });

  test("should mark all fields as edited on markAllEdited", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    act(() => {
      result.current.markAllEdited();
    });

    const { edited } = result.current;

    expect(edited).toEqual({
      name: true,
      age: true,
      isChecked: true,
    });
  });

  test("should call onSuccess when form is valid on handleValidatedSubmit", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    const onSuccess = vi.fn();
    const onError = vi.fn();

    act(() => {
      // 正しいフォーム値をセット
      result.current.handleChange("name", "John");
      result.current.handleChange("age", 25);
      result.current.handleChange("isChecked", true);

      result.current.handleValidatedSubmit(
        onSuccess,
        onError
      )({
        preventDefault: vi.fn(),
      } as unknown as React.FormEvent);
    });

    // 結果を検証
    expect(onSuccess).toHaveBeenCalledWith({
      name: "John",
      age: 25,
      isChecked: true,
    });
    expect(onError).not.toHaveBeenCalled();
  });

  test("should call onError when form is invalid on handleValidatedSubmit", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    const onSuccess = vi.fn();
    const onError = vi.fn();

    act(() => {
      result.current.handleChange("name", "Jo"); // Invalid name
      result.current.handleValidatedSubmit(
        onSuccess,
        onError
      )({ preventDefault: vi.fn() } as any);
    });

    expect(onSuccess).not.toHaveBeenCalled();
    expect(onError).toHaveBeenCalled();
  });
  test("should render error messages correctly for invalid fields", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    act(() => {
      result.current.handleChange("name", "Jo"); // Invalid: Too short
      result.current.handleChange("age", 16); // Invalid: Too young
      result.current.handleChange("isChecked", false); // Invalid: Not checked
    });

    const nameErrorMessages = result.current.renderErrorMessages("name", [
      "Name must be at least 3 characters.",
      "Name must only contain letters.",
    ]);

    const ageErrorMessages = result.current.renderErrorMessages("age", [
      "Age must be greater than 18.",
    ]);

    const isCheckedErrorMessages = result.current.renderErrorMessages(
      "isChecked",
      ["You must check this box."]
    );

    // エラーメッセージのスナップショットテスト
    expect(nameErrorMessages).toMatchSnapshot();
    expect(ageErrorMessages).toMatchSnapshot();
    expect(isCheckedErrorMessages).toMatchSnapshot();
  });

  test("should render error messages with a custom component", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    act(() => {
      result.current.handleChange("name", "Jo"); // Invalid: Too short
    });

    const CustomErrorComponent = ({
      children,
    }: {
      children: React.ReactNode;
    }) => <div className="error">{children}</div>;

    const nameErrorMessages = result.current.renderErrorMessages(
      "name",
      [
        "Name must be at least 3 characters.",
        "Name must only contain letters.",
      ],
      CustomErrorComponent
    );

    expect(nameErrorMessages).toMatchSnapshot();
  });

  test("should render nothing when there are no validation errors", () => {
    const { result } = renderHook(() => useRBoxForm(initialValues, validate));

    act(() => {
      result.current.handleChange("name", "John"); // Valid: Correct name
    });

    const nameErrorMessages = result.current.renderErrorMessages("name", [
      "Name must be at least 3 characters.",
      "Name must only contain letters.",
    ]);

    expect(nameErrorMessages).toMatchSnapshot();
  });
});
