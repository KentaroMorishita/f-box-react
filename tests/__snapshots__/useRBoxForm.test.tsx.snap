// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useRBoxForm > should render error messages correctly for invalid fields 1`] = `
<React.Fragment>
  <span>
    Name must be at least 3 characters.
  </span>
</React.Fragment>
`;

exports[`useRBoxForm > should render error messages correctly for invalid fields 2`] = `
<React.Fragment>
  <span>
    Age must be greater than 18.
  </span>
</React.Fragment>
`;

exports[`useRBoxForm > should render error messages correctly for invalid fields 3`] = `
<React.Fragment>
  <span>
    You must check this box.
  </span>
</React.Fragment>
`;

exports[`useRBoxForm > should render error messages with a custom component 1`] = `
<React.Fragment>
  <CustomErrorComponent>
    Name must be at least 3 characters.
  </CustomErrorComponent>
</React.Fragment>
`;

exports[`useRBoxForm > should render nothing when there are no validation errors 1`] = `<React.Fragment />`;
